### PR TITLE
Migrate deploy to use kne/exec rather than kne/os/exec

### PR DESCRIPTION
**Pull request by bormanp**

### CODE REVIEW

File: deploy/deploy_test.go [hunk 1]
```diff
@@ -68,61 +83,91 @@ func TestKindSpec(t *testing.T) {
...
+		},
 	}, {
 		desc: "create cluster with GAR - failed to get access token",
 		k: &KindSpec{
 			Name:                     "test",
 			GoogleArtifactRegistries: []string{"us-west1-docker.pkg.dev"},
 		},
-		execer:  exec.NewFakeExecer(nil, errors.New("failed to get access token")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}, Err: "failed to get access token"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:126]: this is the cmd that should fail: {Cmd: "gcloud", Args: []string{"auth", "print-access-token"}},

File: deploy/deploy_test.go [hunk 2]
```diff
@@ -68,61 +83,91 @@ func TestKindSpec(t *testing.T) {
...
 	}, {
 		desc: "create cluster with GAR - failed docker login",
 		k: &KindSpec{
 			Name:                     "test",
 			GoogleArtifactRegistries: []string{"us-west1-docker.pkg.dev"},
 		},
-		execer:  exec.NewFakeExecer(nil, nil, errors.New("failed to login to docker")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}},
+			{Cmd: "gcloud", Args: []string{"auth", "print-access-token"}, Err: "failed to login to docker"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:137]: {Cmd: "docker", Args: []string{"login", "-u", "oauth2accesstoken", "-p", "", "https://us-west1-docker.pkg.dev"}},

File: deploy/deploy_test.go [hunk 3]
```diff
@@ -68,61 +83,91 @@ func TestKindSpec(t *testing.T) {
...
 		desc: "create cluster with GAR - failed to get nodes",
 		k: &KindSpec{
 			Name:                     "test",
 			GoogleArtifactRegistries: []string{"us-west1-docker.pkg.dev"},
 		},
-		execer:  exec.NewFakeExecer(nil, nil, nil, errors.New("failed to get nodes")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}},
+			{Cmd: "gcloud", Args: []string{"auth", "print-access-token"}},
+			{Cmd: "docker", Args: []string{"login", "-u", "oauth2accesstoken", "-p", "", "https://us-west1-docker.pkg.dev"}, Err: "failed to get nodes"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:149]: same as the above cases, one error too soon

File: deploy/deploy_test.go [hunk 4]
```diff
@@ -68,61 +83,91 @@ func TestKindSpec(t *testing.T) {
...
 		wantErr: "failed to get nodes",
 	}, {
 		desc: "create cluster with GAR - failed to cp config to node",
 		k: &KindSpec{
 			Name:                     "test",
 			GoogleArtifactRegistries: []string{"us-west1-docker.pkg.dev"},
 		},
-		execer:  exec.NewFakeExecer(nil, nil, nil, nil, errors.New("failed to cp config to node")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}, Err: "failed to cp config to node"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:159]: same

File: deploy/deploy_test.go [hunk 5]
```diff
@@ -68,61 +83,91 @@ func TestKindSpec(t *testing.T) {
...
+		},
 		wantErr: "failed to cp config to node",
 	}, {
 		desc: "create cluster with GAR - failed to restart kubelet",
 		k: &KindSpec{
 			Name:                     "test",
 			GoogleArtifactRegistries: []string{"us-west1-docker.pkg.dev"},
 		},
-		execer:  exec.NewFakeExecer(nil, nil, nil, nil, nil, errors.New("failed to restart kubelet")),
+		resp: []fexec.Response{
```
Comment (alexmasi) [inline deploy/deploy_test.go:168]: same

File: deploy/deploy_test.go [hunk 6]
```diff
@@ -186,7 +255,12 @@ func TestKindSpec(t *testing.T) {
 				"docker": "local",
 			},
 		},
-		execer:  exec.NewFakeExecer(nil, errors.New("unable to pull")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}},
+			{Cmd: "docker", Args: []string{"pull", "docker"}},
```
Comment (alexmasi) [inline deploy/deploy_test.go:260]: i think the err should be here during the pull

File: deploy/deploy_test.go [hunk 7]
```diff
@@ -196,7 +270,12 @@ func TestKindSpec(t *testing.T) {
 				"docker": "local",
 			},
 		},
-		execer:  exec.NewFakeExecer(nil, nil, errors.New("unable to tag")),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}},
+			{Cmd: "docker", Args: []string{"pull", "docker"}},
+			{Cmd: "docker", Args: []string{"tag", "docker", "local"}},
```
Comment (alexmasi) [inline deploy/deploy_test.go:276]: err during tag

File: deploy/deploy_test.go [hunk 8]
```diff
@@ -216,7 +300,9 @@ func TestKindSpec(t *testing.T) {
 				"": "local",
 			},
 		},
-		execer:  exec.NewFakeExecer(nil),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"create", "cluster", "--name", "test"}, Err: "source container must not be empty"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:304]: this should not throw an error, this should actually fail from the non-execed code

File: deploy/deploy_test.go [hunk 9]
```diff
@@ -268,65 +358,82 @@ func TestKindSpec(t *testing.T) {
...
 		wantErr: "failed to get kind version: exec error",
 	}, {
 		desc: "failed kind version parse",
 		k: &KindSpec{
 			Name:    "test",
 			Version: "v0.15.0",
 		},
-		vExecer: exec.NewFakeExecerWithIO(vOut, vOut, exec.Response{Stdout: "kind "}),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"version"}, Stdout: "kind ", Err: "failed to parse kind version from"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:382]: no need to error here, should happen from the result

File: deploy/deploy_test.go [hunk 10]
```diff
@@ -268,65 +358,82 @@ func TestKindSpec(t *testing.T) {
...
 		wantErr: "failed to parse kind version from",
 	}, {
 		desc: "failed kind version got fail parse",
 		k: &KindSpec{
 			Name:    "test",
 			Version: "v0.15.0",
 		},
-		vExecer: exec.NewFakeExecerWithIO(vOut, vOut, exec.Response{Stdout: "kind 0.14.0 go1.18.2 linux/amd64"}),
+		resp: []fexec.Response{
+			{Cmd: "kind", Args: []string{"version"}, Stdout: "kind 0.14.0 go1.18.2 linux/amd64", Err: "kind version check failed: missing prefix on major version"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:392]: same as above

File: deploy/deploy_test.go [hunk 11]
```diff
@@ -416,7 +523,7 @@ func TestMetalLBSpec(t *testing.T) {
 		},
 	}
 	mockCtrl := gomock.NewController(t)
-	defer mockCtrl.Finish()
+	mockCtrl.Finish()
```
Comment (alexmasi) [inline deploy/deploy_test.go:526]: reason for changing this?

File: deploy/deploy_test.go [hunk 12]
```diff
@@ -442,14 +549,18 @@ func TestMetalLBSpec(t *testing.T) {
...
-		execer: exec.NewFakeExecer(errors.New("namespace error")),
-		dErr:   "namespace error",
+		resp: []fexec.Response{
+			{Cmd: "kubectl", Args: []string{"apply", "-f", ""}, Err: "namespace error"},
+		},
+		dErr: "namespace error",
 	}, {
 		desc: "secret create",
+		resp: []fexec.Response{
+			{Cmd: "kubectl", Args: []string{"apply", "-f", ""}, Err: "secret error"},
```
Comment (alexmasi) [inline deploy/deploy_test.go:559]: this shouldn't need to error

File: deploy/deploy_test.go [hunk 13]
```diff
@@ -576,13 +690,18 @@ func TestMetalLBSpec(t *testing.T) {
...
+			{Cmd: "kubectl", Args: []string{"apply", "-f", ""}},
+		},
 		dErr: "dclient error",
 	}, {
 		desc: "valid deployment",
 		m: &MetalLBSpec{
 			IPCount: 20,
 		},
-		execer: exec.NewFakeExecer(nil, nil, nil),
+		resp: []fexec.Response{
```
Comment (alexmasi) [inline deploy/deploy_test.go:702]: why are some of these going from 3 execs to one? was it just a case of the old test having the incorrect amount?

File: deploy/deploy_test.go [hunk 14]
```diff
@@ -24,38 +25,52 @@ import (
 	fakecorev1 "k8s.io/client-go/kubernetes/typed/core/v1/fake"
 
 	ktest "k8s.io/client-go/testing"
+	"k8s.io/klog/v2"
 )
 
+const verbose = true
```
Comment (alexmasi) [inline deploy/deploy_test.go:31]: what is the point of having this as a const? all of the tests use the true value to print their logs

File: deploy/deploy_test.go (hunk 1 above)
Comment (bormanp) [reply deploy/deploy_test.go:126]: Done

File: deploy/deploy_test.go (hunk 14 above)
Comment (bormanp) [reply deploy/deploy_test.go:31]: This should be false normally.  Setting it to true prints out the additional debugging information when working on test (i.e., prints what commands were actually executed).

File: deploy/deploy_test.go (hunk 2 above)
Comment (bormanp) [reply deploy/deploy_test.go:137]: Done

File: deploy/deploy_test.go (hunk 3 above)
Comment (bormanp) [reply deploy/deploy_test.go:149]: Done

File: deploy/deploy_test.go (hunk 5 above)
Comment (bormanp) [reply deploy/deploy_test.go:168]: Done

File: deploy/deploy_test.go (hunk 6 above)
Comment (bormanp) [reply deploy/deploy_test.go:260]: Done

File: deploy/deploy_test.go (hunk 7 above)
Comment (bormanp) [reply deploy/deploy_test.go:276]: Done

File: deploy/deploy_test.go (hunk 8 above)
Comment (bormanp) [reply deploy/deploy_test.go:304]: Done

File: deploy/deploy_test.go (hunk 9 above)
Comment (bormanp) [reply deploy/deploy_test.go:382]: Done

File: deploy/deploy_test.go (hunk 10 above)
Comment (bormanp) [reply deploy/deploy_test.go:392]: Done

File: deploy/deploy_test.go (hunk 11 above)
Comment (bormanp) [reply deploy/deploy_test.go:526]: Nope, that was from when I was trying to figure out why there were mock failures and I failed to change it back.

File: deploy/deploy_test.go (hunk 12 above)
Comment (bormanp) [reply deploy/deploy_test.go:559]: The test hangs without it.  Needs more investigation.

File: deploy/deploy_test.go (hunk 13 above)
Comment (bormanp) [reply deploy/deploy_test.go:702]: There is only one execution during the test.  When I add {Cmd: "x"}, to the list I get
```
    --- FAIL: TestMetalLBSpec/valid_deployment (0.00s)
        deploy_test.go:892: valid deployment: {Cmd: "kubectl", Args: []string{"apply", "-f", ""}}
        deploy_test.go:960: Done: didn't execute:
            	{Cmd: "x"}
```

So yes, I think the old code just had extra nil's in it.

**Review by alexmasi — APPROVED** [review]

**Pull request merged**